cupy/core/_routines_statistics.pyx

```diff
@@ -338,6 +348,50 @@ cdef _mean = create_reduction_func(
      'out0 = a / _type_reduce(_in_ind.size() / _out_ind.size())', None))
 
 
+_count_non_nan = create_reduction_func(
+    'cupy_count_non_nan',
+    ('e->e', 'f->f', 'd->d'),
+    ('(in0 == in0) ? 1.0 : 0.0', 'a + b', 'out0 = a', None), 0)
+
+
+cdef ndarray _nanstd(
+        ndarray a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+    ret = _nanvar(
+        a, axis=axis, dtype=dtype, out=None, ddof=ddof, keepdims=keepdims)
+    return _math._sqrt(ret, dtype=dtype, out=out)
+
+
+cdef ndarray _nanvar(
+        ndarray a, axis=None, dtype=None, out=None, ddof=0, keepdims=False):
+
+    assert a.dtype.kind != 'c', 'Variance for complex numbers is not ' \
+                                'implemented. Current implemention does not ' \
+                                'convert the dtype'
+
+    _count = _count_non_nan(a, axis=axis, dtype=dtype, keepdims=True)
+    arrsum = a._nansum(axis=axis, dtype=dtype, out=None, keepdims=True)
+
+    if out is None:
+        return _nanvar_core(
+            a, arrsum, _count, ddof, axis=axis, keepdims=keepdims)
+    else:
+        return _nanvar_core_out(
+            a, arrsum, _count, ddof, out, axis=axis, keepdims=keepdims)
+
+
+cdef _nanvar_core = ReductionKernel(
+    'S x, T sum, T _count, T ddof', 'T out',
+    '(x == x ? T(pow(x - sum / _count, T(2))) : ' +
+    'T(0)) / max(T(_count - ddof), T(0))',
+    'a + b', 'out = a', '0', '_nanvar_core')
+
+cdef _nanvar_core_out = ReductionKernel(
+    'S x, T sum, T _count, T ddof', 'U out',
+    '(x == x ? T(pow(x - sum / _count, T(2))) : ' +
+    'T(0)) / max(T(_count - ddof), T(0))',
+    'a + b', 'out = a', '0', '_nanvar_core')
+
```

Comment:
How about writing as follows for readability?
```py
cdef _nanvar_preamble = '''
template <typename S, typename T, typename U>
__device__ U nanvar_impl(S x, T mean, long long count) {
    return (x == x ? U((x - mean) * (x - mean)) : U(0)) / count;
}
'''

cdef _nanvar_core = ReductionKernel(
    'S x, T sum, int64 _count, int64 ddof', 'T out',
    'nanvar_impl<S, T, T>(x, sum / _count, max(_count - ddof, 0LL))',
    'a + b', 'out = a', '0', '_nanvar_core', preamble=_nanvar_preamble)

cdef _nanvar_core_out = ReductionKernel(
    'S x, T sum, int64 _count, int64 ddof', 'U out',
    'nanvar_impl<S, T, U>(x, sum / _count, max(_count - ddof, 0LL))',
    'a + b', 'out = a', '0', '_nanvar_core', preamble=_nanvar_preamble)
```

Reply:
Thanks, that looks cleaner. Um, shouldn't it be `T out` in `_nanvar_core`?

Reply:
Yes. `T out` is correct. Thanks.

Reply:
Could you fix test to parameterize also `dtype`?